File: 300. Longest Increasing Subsequence/lengthOfLIS.py
from typing import List

class Solution:
  def lengthOfLIS(self, nums: List[int]) -> int:
    dp = [1] * len(nums)

    for i in range(len(nums) - 1, -1, -1):
      for j in range(i + 1, len(nums)):
        if nums[i] < nums[j]:
          dp[i] = max(dp[i], 1 + dp[j])
  
    return max(dp)
      
      
nums = [10, 9, 2, 5, 3, 7, 101, 18]
solution = Solution()
print('lengthOfLIS: ', solution.lengthOfLIS(nums))

File: 300. Longest Increasing Subsequence/lengthOfLIS.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const lengthOfLIS = (nums) => {
  const dp = Array(nums.length).fill(1);

  for (let i = nums.length - 1; i >= 0; i--) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] < nums[j]) {
        dp[i] = Math.max(dp[i], dp[j] + 1);
      }
    }
  }

  return Math.max(...dp);
};


const nums = [10, 9, 2, 5, 3, 7, 101, 18];
console.log('lengthOfLIS: ', lengthOfLIS(nums)); // 4
